Add multi-material tutorials

Patol75 opened the pull request:
This PR brings multi-material tutorials to the G-ADOPT framework. Happy to take comments and make necessary changes.

drhodrid commented on demos/multi_material/compositional_buoyancy/van_keken_1997.py:
```diff
@@ -0,0 +1,281 @@
+#
+# $$\frac{\partial \psi}{\partial t} + \nabla \cdot \left( \mathbf{u}\psi \right) = 0$$
+#
+# Advection of the level set deteriorates the shape of the initial profile. To maintain
```
I find "deteriorates the shape of the initial profile" a little vague. Can you write explicitly what you mean here?

drhodrid commented on demos/multi_material/compositional_buoyancy/van_keken_1997.py:
```diff
@@ -0,0 +1,281 @@
+    final_checkpoint.save_function(psi, name="Level set")
+# -
+
+# We can visualise the final temperature field using Firedrake's built-in plotting
```
temperature --> level set?

drhodrid commented on demos/multi_material/thermochemical_buoyancy/van_keken_1997.py:
```diff
@@ -0,0 +1,313 @@
+# Rationale
+# ---------
+#
+# One may wish to simulate a geodynamical flow involving multiple physical phases. A
```
For introduction to this tutorial, I think you can be less verbose, building on the previous example.

drhodrid commented on demos/multi_material/thermochemical_buoyancy/van_keken_1997.py:
```diff
@@ -0,0 +1,313 @@
+    final_checkpoint.save_function(psi, name="Level set")
+# -
+
+# We can visualise the final temperature field using Firedrake's built-in plotting
```
Visualise T and phi here?

drhodrid reviewed (commented):
At a glance this looks good. I have made some minor comments and will try running shortly. For my info - this can now be run on master - given everything has been merged?

angus-g commented on demos/multi_material/compositional_buoyancy/van_keken_1997.py:
```diff
@@ -0,0 +1,281 @@
+# time-step length (via a CFL criterion) as the simulation advances in time. We specify
+# the initial time, initial time step $\Delta t$, and output frequency (in time units).
+
+# +
+time_now = 0  # Initial time
+delta_t = Function(R).assign(1)  # Initial time step
+output_frequency = 10  # Frequency (based on simulation time) at which to output
+t_adapt = TimestepAdaptor(
+    delta_t, u, V, target_cfl=0.6, maximum_timestep=output_frequency
+)
+# -
```
```suggestion
time_now = 0  # Initial time
delta_t = Function(R).assign(1)  # Initial time step
output_frequency = 10  # Frequency (based on simulation time) at which to output
t_adapt = TimestepAdaptor(
    delta_t, u, V, target_cfl=0.6, maximum_timestep=output_frequency
)
```

Blocks like this without any blank lines don't need explicit delimiters -- maybe a little easier to read without them?

angus-g commented on demos/multi_material/compositional_buoyancy/van_keken_1997.py:
```diff
@@ -0,0 +1,281 @@
+# This problem has a constant pressure nullspace, handled identically to our previous
+# tutorials.
+
+# +
+Z_nullspace = create_stokes_nullspace(Z)
+# -
```
```suggestion
Z_nullspace = create_stokes_nullspace(Z)
```

angus-g commented on demos/multi_material/compositional_buoyancy/van_keken_1997.py:
```diff
@@ -0,0 +1,281 @@
+# Boundary conditions are specified next: no slip at the top and bottom and free slip
+# on the left and ride sides.
+
+# +
+stokes_bcs = {
+    bottom_id: {"ux": 0, "uy": 0},
+    top_id: {"ux": 0, "uy": 0},
+    left_id: {"ux": 0},
+    right_id: {"ux": 0},
+}
+# -
```
```suggestion
stokes_bcs = {
    bottom_id: {"ux": 0, "uy": 0},
    top_id: {"ux": 0, "uy": 0},
    left_id: {"ux": 0},
    right_id: {"ux": 0},
}
```

drhodrid commented on demos/multi_material/compositional_buoyancy/van_keken_1997.py:
```diff
@@ -0,0 +1,281 @@
+
+# +
+stokes_bcs = {
+    bottom_id: {"ux": 0, "uy": 0},
```
For zero slip, in 3-D example, we set u = 0, so we don't have to set the components individually. Just noting that for consistency.

Patol75 commented:
Thank you for the suggestions; I am working on them.

@drhodrid Yes, everything should run on `master`.

drhodrid commented:
Running them now and will have a few more comments later this eve. For now, all I'd add to previous reviews is:
1. Lovely to see this setup. It actually demonstrates how elegant what you have done is. Well done.
2. Could you call the python files something that makes them easier to distinguish (they currently are both vankekek1997.py).
3. Given that these tutorials are also tested, you'll need to provide some diagnostic values at the end of the run so that @angus-g can add these to our testing platform. The cases run quickly so perhaps increase resolution (x2) so that the tutorials look better resolved and the diagnostic outputs are more meaningful? 
4. In tutorial 1, remove temperature from VTK output, as it is confusing (even though it is zero).

Patol75 commented:
1. I am glad you like it.
2. Given I put them in different subdirectories, I thought that was enough. Or do you prefer I get rid of subdirectories and change file names?
3. Will do. Keep in mind the entrainment diagnostic for the thermochemical case will look nothing like the published benchmark.
4. Well, there will always be a confusion, given that Stokes solver requires a temperature field in the first place.

drhodrid commented on demos/multi_material/compositional_buoyancy/van_keken_1997.py:
```diff
@@ -0,0 +1,276 @@
+V = VectorFunctionSpace(mesh, "CG", 2)  # Velocity function space (vector)
+W = FunctionSpace(mesh, "CG", 1)  # Pressure function space (scalar)
+Z = MixedFunctionSpace([V, W])  # Stokes function space (mixed)
+Q = FunctionSpace(mesh, "CG", 2)  # Temperature function space (scalar)
```
Q not used or needed?

drhodrid commented on demos/multi_material/compositional_buoyancy/van_keken_1997.py:
```diff
@@ -0,0 +1,276 @@
+u, p = split(z)  # Symbolic UFL expressions for velocity and pressure
+z.subfunctions[0].rename("Velocity")  # Associated Firedrake velocity function
+z.subfunctions[1].rename("Pressure")  # Associated Firedrake pressure function
+T = Function(Q, name="Temperature")  # Firedrake function for temperature
```
Does this need to be Q2? Could we not put it on CG 1 (W) function space, thus reducing the need for Q above (given that T is not really necessary for this simulation, aside from G-ADOPT requirement!)

Patol75 replied:
That will not change the fact that `T` is there. Defining `Q` aligns with other tutorials.

drhodrid replied:
True - but it's not really used here, hence confusing. Happy to leave as is for now and come back to this down the track if we find a way to not require T for simple multi-material Stokes solves.

drhodrid commented on demos/multi_material/compositional_buoyancy/van_keken_1997.py:
```diff
@@ -0,0 +1,276 @@
+
+# +
+import numpy as np
+import shapely as sl
```
Just wanted to confirm that shapely was now included as a g-adopt dependency? When I installed g-adopt using pip install -e earlier (after the merge of level sets), I still needed to install shapely via pip. Is this the expected behavior or should it now come in via g-adopt?

Patol75 replied:
What @angus-g said. :)

drhodrid commented on demos/multi_material/compositional_buoyancy/van_keken_1997.py:
```diff
@@ -0,0 +1,276 @@
+isd_params = (interface_deflection, interface_wavelength, material_interface_y)
+
+# Shapely LineString representation of the material interface
+interface_x = np.linspace(0, lx, 1000)
```
Out of interest, are results at all sensitivity to the choice of 1000 here? Obviously irrelevant for this tutorial, but perhaps when we think about publishing this approach. 

Probably a comment here too about why you chose 1000?

Patol75 replied:
This number defines the amount of points used to discretise the initial interface. You just need a number that is sufficiently large to capture the wiggles, if any, of the initial interface. If you put something too low, your initial state will not look like what you expected on ParaView.

drhodrid replied:
Does does the discretised interface ultimately smear to some extent, and this 1000 is just to ensure the best possible representation of that initial condition with our current discretisation?

drhodrid commented on demos/multi_material/compositional_buoyancy/van_keken_1997.py:
```diff
@@ -0,0 +1,276 @@
+Ra = 0  # Thermal Rayleigh number
+
+RaB = field_interface(
+    [psi], [material.RaB for material in materials], method="arithmetic"
```
Worth adding a comment here about the "arithmetic" averaging approach used at the interface (I assume that's what this is?) alongside the equation? Do you perhaps want to quickly note the other options available (geometric, harmonic) and those equations too?

drhodrid commented on demos/multi_material/compositional_buoyancy/van_keken_1997.py:
```diff
@@ -0,0 +1,276 @@
+# This problem has a constant pressure nullspace, handled identically to our previous
+# tutorials.
+
+Z_nullspace = create_stokes_nullspace(Z)
```
I realise it does the same thing, but for consistency with previous tutorials, do you want to say closed = True, rotational = False?

Patol75 replied:
I edited the comment above.

drhodrid commented on demos/multi_material/compositional_buoyancy/van_keken_1997.py:
```diff
@@ -0,0 +1,276 @@
+
+Z_nullspace = create_stokes_nullspace(Z)
+
+# Boundary conditions are specified next: no slip at the top and bottom and free slip
```
Is it worth adding a comment here that since all domain boundaries are closed, there is no boundary condition for the level set required?

drhodrid commented on demos/multi_material/compositional_buoyancy/van_keken_1997.py:
```diff
@@ -0,0 +1,276 @@
+plog.log_str("step time dt u_rms entrainment")
+
+gd = GeodynamicalDiagnostics(z, T, bottom_id, top_id)
+material_area = material_interface_y * lx  # Quantity of tracked material
```
Instead of quantity - I would say areal fraction, or volume fraction? It looks like that's what this is?

Patol75 replied:
It is not a fraction though. It is the area occupied by the material in axis units. It would be a volume in 3-D.

drhodrid replied:
Just say area then? "Quantity" is a bit vague.

drhodrid commented on demos/multi_material/compositional_buoyancy/van_keken_1997.py:
```diff
@@ -0,0 +1,276 @@
+
+gd = GeodynamicalDiagnostics(z, T, bottom_id, top_id)
+material_area = material_interface_y * lx  # Quantity of tracked material
+entrainment_height = 0.2  # Height above which entrainment diagnostic is calculated
```
Say a little more about what this diagnostic is (i.e. the integral of a certain material above the specified height)?

drhodrid commented on demos/multi_material/compositional_buoyancy/van_keken_1997.py:
```diff
@@ -0,0 +1,276 @@
+    transpose_nullspace=Z_nullspace,
+)
+
+subcycles = 1  # Number of advection solves to perform within one time step
```
Provide some context for this? You use 1, but is this the recommendation for users too? What are the other options available and what are the trade-offs?

Patol75 replied:
This is used when the CFL for the energy system differs from that of the level-set advection. The latter should not exceed 0.6, but @stephankramer mentioned that the one for the energy system is usually higher. Subcycling allows multiple level-set advection steps without updating Stokes and energy in the meantime.

drhodrid commented on demos/multi_material/compositional_buoyancy/van_keken_1997.py:
```diff
@@ -0,0 +1,276 @@
+
+subcycles = 1  # Number of advection solves to perform within one time step
+level_set_solver = LevelSetSolver(psi, u, delta_t, eSSPRKs10p3, subcycles, epsilon)
+# Increase the reinitialisation time step to make up for the coarseness of the mesh
```
Make it cleaar that this is increased relative to the default values. If you increase the resolution of this case, will you need to revisit this parameter choice? How should a user pick the value?

Patol75 replied:
I think it is pretty clear. With increased resolution, I would expect this parameter not to require as much of an increase. I do not really have clever advice on how to choose ideal reinitialisation parameters. It is a little bit of trial and error.

drhodrid commented on demos/multi_material/thermochemical_buoyancy/van_keken_1997.py:
```diff
@@ -0,0 +1,277 @@
+# convection presented in van Keken et al. (1997). Inside a 2-D domain heated from
+# below, a denser material sits at the bottom boundary beneath a lighter material.
+# Whilst the compositional stratification is stable, heat transfer from the boundary
+# generates positive buoyancy in the denser material, allowing thin layers of material
```
Instead of thing layers, perhaps use thin "tendrils" - that's the lingo I seem to remember.

drhodrid replied:
I would add a comment at the end of this paragraph that the goal of this tutorial is to illustrate how to set up such a problem. To make the tutorial tractable, it must be run at a coarser resolution than is required, and hence results differ to those in the benchmark. Nonetheless, you have verified, by simulating this case at higher resolution, that things do converge towards those benchmark values.

drhodrid commented on demos/multi_material/thermochemical_buoyancy/van_keken_1997.py:
```diff
@@ -0,0 +1,277 @@
+}
+temp_bcs = {bottom_id: {"T": 1}, top_id: {"T": 0}}
+
+# We move on to initialising the temperature field.
```
For consistency with other tutorials, could the temperature field be initialised before boundary conditions are set for T? I realise it doesn't make a difference, but the more consistent we are for a user, the better.

Patol75 replied:
I need to impose the boundary conditions on the initial state; that is why I swapped the order.

drhodrid reviewed (commented):
A few more specific comments. I really enjoyed going through this and learning. Goes to show the benefit of a nice tutorial! Well done!

drhodrid commented on demos/multi_material/thermochemical_buoyancy/van_keken_1997.py:
```diff
@@ -0,0 +1,277 @@
+# +
+step = 0  # A counter to keep track of looping
+output_counter = 0  # A counter to keep track of outputting
+time_end = 0.05
```
Although the benchmark goes for this time, given we are under resolved, perhaps we only need to run the tutorial to 0.02 or similar (and we can flag that the true benchmark runs for longer with a comment)?

Patol75 commented:
@drhodrid I have addressed a few comments and added responses and reactions to others. More from me tomorrow.

drhodrid commented:
Roger that. It’s looking good!

On 4 Jun 2024, at 8:27 PM, Thomas Duvernay ***@***.***> wrote:



@drhodrid<https://github.com/drhodrid> I have addressed a few comments and added responses and reactions to others. More from me tomorrow.

—
Reply to this email directly, view it on GitHub<https://github.com/g-adopt/g-adopt/pull/79#issuecomment-2147178191>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AB25UKSG5Q7M3HQPUVZO5YDZFWJBBAVCNFSM6AAAAABIXYVVWCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCNBXGE3TQMJZGE>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

Patol75 commented:
Today's update:

- **Diagnostics** Can I just provide the RMS velocity and entrainment values at the end of each simulation?
- I have addressed all comments.
- I am keen on seeing the rendered version. That will allow me to make further targeted changes.

angus-g commented:
>     * I am keen on seeing the rendered version. That will allow me to make further targeted changes.

https://78ff9a48.g-adopt.pages.dev/tutorials/compositional_buoyancy/ and https://78ff9a48.g-adopt.pages.dev/tutorials/thermochemical_buoyancy/ -- I guess first comment is to give them different titles!

Patol75 commented:
Thank you, Angus. New changes pushed.

angus-g commented on demos/multi_material/compositional_buoyancy/compositional_buoyancy.py:
```diff
@@ -0,0 +1,291 @@
+# -
+
+# We now provide initial conditions for the level-set field. To this end, we use the
+# `shapely`` library to represent the initial location of the material interface and
```
Typo?

```suggestion
# `shapely` library to represent the initial location of the material interface and
```

Patol75 replied:
Good catch. I blame that one on VSCode.

angus-g reviewed (approved):
Looks good from my end. Will let @drhodrid chime in with any extra comments when he gets a chance to take a look.

drhodrid commented:
I'd say these look really great and are pretty much good to go. I only have two minor comments:
1. For both cases, explain that resolving the material interface typically requires resolution beyond that provided in these tutorials. The goal of the tutorials is merely to showcase how things are set up.
2. For the thermo-chemical tutorial, where you now use gmsh (a nice addition which we don't have in our other tutorials!), I think that gmsh code block should be separated from the function spaces etc that follow.... I would also add a few more comments explaining what the key lines are doing for the gmsh generation.

Patol75 commented:
@drhodrid For your first point, I do not think it applies to the compositional benchmark — the result is pretty convincing if you ask me — and I believe I already mentioned it in the introduction to the thermochemical benchmark. For your second point, I separated the GMSH block from the rest, but I have not added comments to the part where the GMSH API is used. Two reasons: the GMSH API is pretty self-explanatory (when you choose appropriate names for variables, functions, methods, and classes, usually you do not need many comments) and this should not be a GMSH tutorial.

drhodrid commented:
Ahhh I hadn't seen that change to the second tutorial. Nicely done. 

And yes ok - furry muff regarding gmsh. This is now good to go IMO. Did you want to merge?

Patol75 commented:
@angus-g Do I need the regression test here? I cancelled it...
